### PR TITLE
[ads] Transition text classification v2/v1 models from JSON to FlatBuffers

### PR DESCRIPTION
**Pull request by aseren**
Current PR does the following:
* Adds support of new v2 (neural) FlatBuffers text classification model
* Adds support of old v1 (linear) text classification model converted to FlatBuffers (the previous format was JSON)
* Tries to load v2 (neural) text classification model first
* Fallbacks to the old v1 (linear) text classification model, if the v2 (neural) model failed

Resolves https://github.com/brave/brave-browser/issues/33336

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Test case 1
* Set English language system locale
* Run the browser with a fresh profile
* Join Brave Rewards
* Make sure that Brave Ads resource for EN language was downloaded on brave://components
* Make sure that the v2 (neural) text classification resource is loaded and initialized. There should be a line in the brave rewards log: `Successfully initialized awdrtoxdlcdyjmckgmtqptnhdtlxgblp neural text classification resource`
* Visit a web page. Make sure that the page is classified. For example visiting `https://bitcoin.org/en/` gives the following line in brave rewards log: `Classified text with the top segment as crypto-crypto`
* Make sure that notification ads can be served.

### Test case 2
* Set non-English language system locale
* Run the browser with a fresh profile
* Join Brave Rewards
* Make sure that Brave Ads resource for language was downloaded on brave://components
* Make sure that the v1 (linear) text classification resource is loaded and initialized. There should be a line in the brave rewards log: `Successfully initialized awdrtoxdlcdyjmckgmtqptnhdtlxgblp linear text classification resource`
* Visit a web page. Make sure that the page is classified.
* Make sure that notification ads can be served.

### Test case 3
* Set English language system locale
* Run the browser with a fresh profile
* Join Brave Rewards
* Make sure that Brave Ads resource for EN language was downloaded on brave://components
* Exit Brave Browser
* Open Brave Ads resource component directory at `Brave-Browser/kglppciafpehhjdhgefiljcmemakfjhk/{version}`
* Delete `svymudlfwueobbiprqazdzkdeslimohu` file
* Create `svymudlfwueobbiprqazdzkdeslimohu` file with some random text
* Run the browser
* Make sure that the text classification resource isn't loaded. There should be a line in the brave rewards log: `Failed to initialize awdrtoxdlcdyjmckgmtqptnhdtlxgblp text classification resource (Failed to load flatbuffers text classification pipeline)`
* Make sure that notification ads can be served.

### CODE REVIEW

File: components/brave_ads/core/internal/common/resources/flat/text_classification_neural_model.fbs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+table Classifier {
+  classes: [string] (required);
+  classifier_type: string (required);
+  neural_matricies_data: [NeuralMatrixData] (required);
+  neural_post_matrix_functions: [string] (required);
+}
+
+table NeuralModel {
+  classifier: Classifier (required);
+  locale: string (required);
```
Comment (tmancey) [inline components/brave_ads/core/internal/common/resources/flat/text_classification_neural_model.fbs:22]: @ptjames @aseren do we validate that the users current language matches during classification?

File: components/brave_ads/core/internal/common/resources/flatbuffers_resources_util_impl.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  const base::ScopedFILE scoped_file(base::FileToFILE(std::move(file), "rb"));
+  if (!base::ReadStreamToString(scoped_file.get(), &buffer)) {
+    return base::unexpected("Failed to read file");
+  }
+
+  return T::CreateFromFlatBuffers(std::move(buffer));
+}
+
+template <typename T>
+void LoadFileFlatBuffersResourceCallback(
```
Comment (tmancey) [inline components/brave_ads/core/internal/common/resources/flatbuffers_resources_util_impl.h:39]: LoadFlatBuffersResourceFileCallback

File: components/brave_ads/core/internal/ml/data/vector_data.cc [hunk 3]
```diff
@@ -176,6 +180,12 @@ void VectorData::DivideByScalar(const float scalar) {
   }
 }
 
+float VectorData::GetSum() const {
+  return static_cast<float>(std::accumulate(
+      storage_->values().cbegin(), storage_->values().cend(), 0.0,
+      [](const float& lhs, const float rhs) -> float { return lhs + rhs; }));
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/data/vector_data.cc:186]: can we remove & for lhs

File: components/brave_ads/core/internal/ml/data/vector_data.cc [hunk 4]
```diff
@@ -184,6 +194,30 @@ float VectorData::GetNorm() const {
...
+}
+
+void VectorData::Softmax() {
+  float maximum = -std::numeric_limits<float>::infinity();
+  for (float& value : storage_->values()) {
+    maximum = (value > maximum) ? value : maximum;
+  }
+  float sum_exp = 0.0;
+  for (float& value : storage_->values()) {
+    const float value_exp = std::exp(value - maximum);
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/data/vector_data.cc:213]: nit: could we not remove this local variable, and just `+ std::exp...`?

File: components/brave_ads/core/internal/ml/model/neural/neural_unittest.cc [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+      BuildNeuralModel(matricies, post_matrix_functions, classes));
+  ASSERT_TRUE(neural);
+  const VectorData sample_observation({0.2, 0.65, 0.15});
+
+  // Act
+  const absl::optional<PredictionMap> sample_predictions =
+      neural->Predict(sample_observation);
+
+  // Assert
+  ASSERT_TRUE(sample_predictions);
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/model/neural/neural_unittest.cc:128]: nit: could we move this to just after the usage

File: components/brave_ads/core/internal/ml/model/neural/neural_unittest.cc [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/model/neural/neural_unittest.cc:156]: nit: could we move this to just after the usage

File: components/brave_ads/core/internal/ml/model/neural/neural_unittest.cc [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/model/neural/neural_unittest.cc:189]: nit: could we move this to just after the usage

File: components/brave_ads/core/internal/ml/model/neural/neural_unittest.cc [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+
+  // Act
+  const absl::optional<PredictionMap> sample_predictions =
+      neural->GetTopPredictions(sample_observation);
+  const absl::optional<PredictionMap> sample_predictions_constrained =
+      neural->GetTopCountPredictions(sample_observation, 2);
+
+  // Assert
+  ASSERT_TRUE(sample_predictions);
+  ASSERT_TRUE(sample_predictions_constrained);
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/model/neural/neural_unittest.cc:225]: nit: could we move these to just after the usage

File: components/brave_ads/core/internal/ml/transformation/distribution_transformation_unittest.cc [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+  // // Act
+  data = to_distribution.Apply(data);
+  const VectorData* const transformed_vector_data =
+      static_cast<VectorData*>(data.get());
+  const std::vector<float> transformed_vector_values =
+      transformed_vector_data->GetData();
+
+  // Assert
+  ASSERT_EQ(DataType::kVector, data->GetType());
+  ASSERT_EQ(transformed_vector_values.size(), 5U);
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/transformation/distribution_transformation_unittest.cc:39]: nit: could we move this to just after the usage

File: components/brave_ads/core/internal/ml/transformation/distribution_transformation_unittest.cc [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/transformation/distribution_transformation_unittest.cc:66]: nit: could we move this to just after the usage

File: components/brave_ads/core/internal/ml/transformation/distribution_transformation_unittest.cc [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+  // // Act
+  data = to_distribution.Apply(data);
+  const VectorData* const transformed_vector_data =
+      static_cast<VectorData*>(data.get());
+  const std::vector<float> transformed_vector_values =
+      transformed_vector_data->GetData();
+
+  // Assert
+  ASSERT_EQ(DataType::kVector, data->GetType());
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/transformation/distribution_transformation_unittest.cc:89]: nit: could we move this to just after the usage

File: components/brave_ads/core/internal/ml/transformation/distribution_transformation_unittest.cc [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+  constexpr char kTestString[] = "this is a test string";
+  std::unique_ptr<Data> data = std::make_unique<TextData>(kTestString);
+
+  const DistributionTransformation to_distribution;
+
+  // // Act
+  data = to_distribution.Apply(data);
+
+  // Assert
+  EXPECT_TRUE(data == nullptr);
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/transformation/distribution_transformation_unittest.cc:104]: EXPECT_FALSE(data);

File: components/brave_ads/core/internal/ml/transformation/mapped_tokens_transformation_unittest.cc [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+
+  std::vector<float> transformed_vector_values(
+      transformed_vector_data->GetDimensionCount());
+  transformed_vector_values =
+      transformed_vector_data->GetData(transformed_vector_values);
+
+  // Assert
+  ASSERT_EQ(DataType::kVector, data->GetType());
+  ASSERT_TRUE(transformed_vector_values.size() ==
+              static_cast<size_t>(vector_dimension));
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/transformation/mapped_tokens_transformation_unittest.cc:135]: nit: could we move this to just after the usage

File: components/brave_ads/core/internal/targeting/contextual/text_classification/resource/text_classification_resource.cc [hunk 14]
```diff
@@ -52,33 +57,33 @@ void TextClassificationResource::MaybeLoadOrReset() {
 void TextClassificationResource::Load() {
   did_load_ = true;
 
-  LoadAndParseResource(kTextClassificationResourceId,
-                       kTextClassificationResourceVersion.Get(),
-                       base::BindOnce(&TextClassificationResource::LoadCallback,
-                                      weak_factory_.GetWeakPtr()));
+  LoadFlatBuffersResource(
+      kFlatBuffersTextClassificationResourceId,
+      kFlatBuffersTextClassificationResourceVersion.Get(),
```
Comment (tmancey) [inline components/brave_ads/core/internal/targeting/contextual/text_classification/resource/text_classification_resource.cc:62]: @aseren is this related to the versioning you removed?

File: components/brave_ads/core/internal/ml/pipeline/text_processing/text_processing.cc [hunk 15]
```diff
@@ -68,58 +78,94 @@ bool TextProcessing::SetPipeline(base::Value::Dict dict) {
...
+    return neural_model_->GetTopPredictions(*vector_data);
+  }
+  return absl::nullopt;
+}
+
 absl::optional<PredictionMap> TextProcessing::Apply(
     std::unique_ptr<Data> input_data) const {
-  std::unique_ptr<Data> current_data = std::move(input_data);
-  CHECK(current_data);
+  std::unique_ptr<Data> mutable_input_data = std::move(input_data);
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/pipeline/text_processing/text_processing.cc:111]: input_data is already mutable

File: components/brave_ads/core/internal/ml/pipeline/text_processing/text_processing.cc [hunk 16]
```diff
@@ -68,58 +78,94 @@ bool TextProcessing::SetPipeline(base::Value::Dict dict) {
...
+}
+
+absl::optional<PredictionMap> TextProcessing::GetTopPredictions(
+    const std::string& text) const {
+  const absl::optional<PredictionMap> predictions = GetPredictions(text);
+  if (!predictions) {
+    return absl::nullopt;
+  }
+  PredictionMap top_predictions = FilterPredictions(*predictions);
+  return top_predictions;
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/pipeline/text_processing/text_processing.cc:161]: we could just `return FilterPredictions...`

File: components/brave_ads/core/internal/common/resources/flat/text_classification_neural_model.fbs (hunk 1 above)
Comment (ptjames) [reply components/brave_ads/core/internal/common/resources/flat/text_classification_neural_model.fbs:22]: If a user changes system language, is the new language component resource not pulled (if not already available locally)? If not, that seems like it should be general behavior

Comment (tmancey) [reply components/brave_ads/core/internal/common/resources/flat/text_classification_neural_model.fbs:22]: Yes they are but only if we have a supported language, but we should check that the resource code resets the resource for non supported locales so that we do not classify in those scenarios.

File: components/brave_ads/core/internal/targeting/contextual/text_classification/resource/text_classification_resource.cc (hunk 14 above)
Comment (aseren) [reply components/brave_ads/core/internal/targeting/contextual/text_classification/resource/text_classification_resource.cc:62]: No, this is the resource version from resource.json file

File: components/brave_ads/core/internal/ml/pipeline/text_processing/text_processing.cc (hunk 15 above)
Comment (aseren) [reply components/brave_ads/core/internal/ml/pipeline/text_processing/text_processing.cc:111]: This is fine here to have a local variable because it is re-assigned to another object later:
```
  const size_t transformation_count = transformations_.size();
  for (size_t i = 0; i < transformation_count; ++i) {
    mutable_input_data = transformations_[i]->Apply(mutable_input_data);
```

File: components/brave_ads/core/internal/common/resources/flat/text_classification_neural_model.fbs (hunk 1 above)
Comment (aseren) [reply components/brave_ads/core/internal/common/resources/flat/text_classification_neural_model.fbs:21]: added language code checking

File: components/brave_ads/core/internal/ml/data/vector_data.cc [hunk 17]
```diff
@@ -219,4 +258,19 @@ const std::vector<float>& VectorData::GetData() const {
...
+    std::vector<float>& dense_vector) const {
+  size_t dimension_count = GetDimensionCount();
+  if (storage_->values().size() == dimension_count) {
+    return storage_->values();
+  }
+  if (dense_vector.size() != dimension_count) {
+    return storage_->values();
+  }
+  for (size_t i = 0; i < storage_->points().size(); i++) {
+    dense_vector[storage_->points()[i]] = storage_->values()[i];
```
Comment (aseren) [inline components/brave_ads/core/internal/ml/data/vector_data.cc:271]: This can lead to out-of-bounds access if dense_vector is not properly constructed (which is actually done in neural.cc). Need to rewrite this function.

Comment (aseren) [reply components/brave_ads/core/internal/ml/data/vector_data.cc:269]: Changed to `std::vector<float> VectorData::GetDenseData();`

**Review by kylehickinson — APPROVED** [review]

**Review by tmancey — APPROVED** [review]
LGTM++

**Review by ptjames — APPROVED** [review]

File: components/brave_ads/core/internal/targeting/contextual/text_classification/resource/text_classification_resource.cc [hunk 18]
```diff
@@ -37,6 +45,20 @@ TextClassificationResource::~TextClassificationResource() {
   RemoveAdsClientNotifierObserver(this);
 }
 
+void TextClassificationResource::ClassifyPage(const std::string& text,
+                                              ClassifyPageCallback callback) {
+  if (!IsInitialized()) {
+    BLOG(1,
+         "Failed to process text classification as resource not initialized");
+    return std::move(callback).Run({});
```
Comment (tmancey) [inline components/brave_ads/core/internal/targeting/contextual/text_classification/resource/text_classification_resource.cc:53]: nit: comment for `{}`, i.e. `/*probabilities=*/{}`

File: components/brave_ads/core/internal/targeting/contextual/text_classification/resource/text_classification_resource.cc [hunk 19]
```diff
@@ -52,32 +74,41 @@ void TextClassificationResource::MaybeLoadOrReset() {
...
+      GetResourceId(), GetResourceVersion(),
+      base::BindOnce(&TextClassificationResource::LoadFileResourceCallback,
+                     weak_factory_.GetWeakPtr()));
 }
 
-void TextClassificationResource::LoadCallback(
-    ResourceParsingErrorOr<ml::pipeline::TextProcessing> result) {
+void TextClassificationResource::LoadFileResourceCallback(base::File file) {
+  if (!file.IsValid()) {
+    return;
```
Comment (tmancey) [inline components/brave_ads/core/internal/targeting/contextual/text_classification/resource/text_classification_resource.cc:85]: nit: log file is invalid

File: components/brave_ads/core/internal/targeting/contextual/text_classification/resource/text_classification_resource.cc [hunk 20]
```diff
@@ -52,32 +74,41 @@ void TextClassificationResource::MaybeLoadOrReset() {
...
                                            << result.error() << ")");
   }
 
-  if (!result.value().IsInitialized()) {
-    return BLOG(1, kTextClassificationResourceId
-                       << " text classification resource is not available");
-  }
-
-  BLOG(1, "Successfully loaded " << kTextClassificationResourceId
+  const std::string pipeline_type = result.value() ? "neural" : "linear";
```
Comment (tmancey) [inline components/brave_ads/core/internal/targeting/contextual/text_classification/resource/text_classification_resource.cc:106]: nit: `const bool is_neural = result.value()`

File: components/brave_ads/core/internal/targeting/contextual/text_classification/text_classification_feature.h [hunk 21]
```diff
@@ -16,6 +16,9 @@ BASE_DECLARE_FEATURE(kTextClassificationFeature);
 constexpr base::FeatureParam<int> kTextClassificationResourceVersion{
     &kTextClassificationFeature, "resource_version", 1};
```
Comment (tmancey) [inline components/brave_ads/core/internal/targeting/contextual/text_classification/text_classification_feature.h:17]: Discussed with @aseren and we can remove this legacy feature

File: components/brave_ads/core/internal/targeting/contextual/text_classification/text_classification_processor.cc [hunk 22]
```diff
@@ -65,10 +66,13 @@ void TextClassificationProcessor::Process(const std::string& text) {
     return BLOG(1, "Text not classified as not enough content");
   }
 
-  const std::string segment =
+  for (const auto& [segment, probability] : *probabilities) {
+    BLOG(6, segment << ": " << probability);
+  }
```
Comment (tmancey) [inline components/brave_ads/core/internal/targeting/contextual/text_classification/text_classification_processor.cc:71]: Discussed with @aseren and @ptjames to remove, thanks

File: components/brave_ads/core/test/data/ml/pipeline/text_processing/not_valid_model [hunk 23]
```diff
@@ -0,0 +1 @@
+not_valid_model
```
Comment (tmancey) [inline components/brave_ads/core/test/data/ml/pipeline/text_processing/not_valid_model:1]: nit: `invalid_model` for filename

File: components/brave_ads/core/internal/ml/transformation/mapped_tokens_transformation.h [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+
+#include "base/memory/raw_ptr.h"
+#include "brave/components/brave_ads/core/internal/ml/transformation/transformation.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+namespace brave_ads::neural_text_classification::flat {
+struct MappedTokenTransformation;
+}  // namespace brave_ads::neural_text_classification::flat
+
+namespace brave_ads::ml {
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/transformation/mapped_tokens_transformation.h:22]: nit: ```
namespace brave_ads {

namespace neural_text_classification::flat {
struct MappedTokenTransformation;
}  // namespace neural_text_classification::flat

namespace ml {
```

Comment (tmancey) [reply components/brave_ads/core/internal/ml/transformation/mapped_tokens_transformation.h:22]: Maybe a followup if you agree?

File: components/brave_ads/core/internal/ml/transformation/mapped_tokens_transformation_unittest.cc [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+  const VectorData* const transformed_vector_data =
+      static_cast<VectorData*>(data.get());
+
+  const std::vector<float> transformed_vector_values =
+      transformed_vector_data->GetDenseData();
+  ASSERT_TRUE(transformed_vector_values.size() ==
+              static_cast<size_t>(vector_dimension));
+
+  // Assert
+  EXPECT_TRUE((std::fabs(1.0 - transformed_vector_values.at(0)) < kTolerance) &&
```
Comment (tmancey) [inline components/brave_ads/core/internal/ml/transformation/mapped_tokens_transformation_unittest.cc:94]: nit: remove surrounding parenthesis for each condition

File: components/brave_ads/core/test/data/ml/pipeline/text_processing/not_valid_model (hunk 23 above)
Comment (tmancey) [reply components/brave_ads/core/test/data/ml/pipeline/text_processing/not_valid_model:1]: nit: only_required_fields_neural_model

**Review by tmancey — APPROVED** [review]
LGTM++ (left some nits)
